### PR TITLE
[3D] 3d tilt

### PR DESCRIPTION
**Pull request by fredj**

### CODE REVIEW

**Comment by fredj** [comment]
First look: https://mf-geoadmin3.dev.bgdi.ch/3d_tilt/?dev3d=true

**Comment by fredj** [comment]
demo updated

**Comment by fredj** [comment]
Thanks for any review.

I don't really like the new `disable3d` property for the swissimage layer. Any other proposal welcome !

File: src/components/tilt3d/Tilt3dDirective.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      templateUrl: 'components/tilt3d/partials/tilt3d.html',
+      link: function(scope, element, attrs) {
+
+        scope.supported = gaBrowserSniffer.webgl;
+
+        // FIXME: remove this
+        // true is the selected background layer is not 3d compatible.
+        scope.disabled = false;
+        scope.$on('gaBgChange', function(evt, value) {
+          scope.disabled = value.id == 'ch.swisstopo.swissimage';
```
Comment (elemoine) [inline src/components/tilt3d/Tilt3dDirective.js]: Why don't you use `value.disable3d` here?

File: src/index.mako.html [hunk 2]
```diff
@@ -217,6 +217,7 @@
   % if device == 'mobile':
         <div ga-offline-bt></div>
   % endif
+        <div ga-tilt3d ng-if="globals.dev3d"></div>
```
Comment (elemoine) [inline src/index.mako.html]: Don't you think that using `ng-show` would be a better choice here?

**Comment by elemoine** [comment]
Question for everyone: shouldn't the mouse cursor change when hovering the 3D button and when 3D is disabled because the background layer is swissimage? Currently, the cursor is still a hand, which I think is confusing.

**Comment by elemoine** [comment]
@fredj, when I switch to 3D I sometimes get some big jump with a blue background before actually seeing the expected background. Is this a known issue?

File: src/components/tilt3d/Tilt3dDirective.js (hunk 1 above)
Comment (fredj) [inline src/components/tilt3d/Tilt3dDirective.js]: `disable3d` property is only available in the background selector directive, not here

**Comment by elemoine** [comment]
> @fredj, when I switch to 3D I sometimes get some big jump with a blue background before actually seeing the expected background. Is this a known issue?

And when this occurs the map is not tilted after switching to 3D. I am not even sure the location is the same as the previous, 2D, location.

File: src/index.mako.html (hunk 2 above)
Comment (fredj) [inline src/index.mako.html]: The advantage of `ng-if` is that the directive is not initialized at all if the condition is false. (but I'm not opposed to change it)

**Comment by fredj** [comment]
> @fredj, when I switch to 3D I sometimes get some big jump with a blue background before actually seeing the expected background. Is this a known issue?

yes, it's a known ol3-cesium issue

**Comment by fredj** [comment]
> Question for everyone: shouldn't the mouse cursor change when hovering the 3D button and when 3D is disabled because the background layer is swissimage? Currently, the cursor is still a hand, which I think is confusing.

This is correct. The UI will by done by swisstopo.

File: src/components/tilt3d/Tilt3dDirective.js (hunk 1 above)
Comment (oterral) [inline src/components/tilt3d/Tilt3dDirective.js]: it shouldn't, The `value`param should contains the `disable3d`property.

**Comment by fredj** [comment]
I've fixed the css cursor when the feature is disabled or not available

File: src/components/tilt3d/Tilt3dDirective.js (hunk 1 above)
Comment (fredj) [inline src/components/tilt3d/Tilt3dDirective.js]: oups, I was searching in the `evt` object ...

Comment (fredj) [inline src/components/tilt3d/Tilt3dDirective.js]: fixed, code updated

File: src/components/tilt3d/Tilt3dDirective.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      restrict: 'A',
+      templateUrl: 'components/tilt3d/partials/tilt3d.html',
+      link: function(scope, element, attrs) {
+
+        scope.supported = gaBrowserSniffer.webgl;
+
+        // true is the selected background layer is not 3d compatible.
+        scope.disabled = false;
+        scope.$on('gaBgChange', function(evt, value) {
+          scope.disabled = !!value.disable3d;
```
Comment (elemoine) [inline src/components/tilt3d/Tilt3dDirective.js]: Why `!!`? `value.disable3d` is a boolean value already, no?

File: src/index.mako.html (hunk 2 above)
Comment (elemoine) [inline src/index.mako.html]: Ok, It may be quite the same here. Personally I tend to use `ng-if` only when strictly necessary.

**Comment by elemoine** [comment]
No more comments from me.

File: src/components/tilt3d/Tilt3dDirective.js (hunk 3 above)
Comment (fredj) [inline src/components/tilt3d/Tilt3dDirective.js]: `boolean` or `undefned`

Comment (elemoine) [inline src/components/tilt3d/Tilt3dDirective.js]: Oh yes, true.

**Comment by gjn** [comment]
LGTM. Artwork still remains to be done. But I'll merge this.

Thanks alot.

**Pull request merged**